channeldb/migrations.go

```diff
@@ -458,3 +459,107 @@ func paymentStatusesMigration(tx *bolt.Tx) error {
+// to a miscalculation on the offset of an edge's policy last update. This
+// migration also fixes the case where the public keys within edge policies were
+// being serialized with an extra byte, causing an even greater error when
+// attempting to perform the offset calculation described earlier.
+func migratePruneEdgeUpdateIndex(tx *bolt.Tx) error {
+	// To begin the migration, we'll retrieve the update index bucket. If it
+	// does not exist, we have nothing left to do so we can simply exit.
+	edges := tx.Bucket(edgeBucket)
+	if edges == nil {
+		return nil
+	}
+	edgeUpdateIndex := edges.Bucket(edgeUpdateIndexBucket)
+	if edgeUpdateIndex == nil {
+		return nil
+	}
+
+	// Retrieve some buckets that will be needed later on. These should
+	// already exist given the assumption that the buckets above do as well.
+	edgeIndex := edges.Bucket(edgeIndexBucket)
+	if edgeIndex == nil {
+		return errors.New("edge index should exist but does not")
+	}
+	nodes := tx.Bucket(nodeBucket)
+	if nodes == nil {
+		return errors.New("node bucket should exist but does not")
+	}
+
+	log.Info("Migrating database to properly prune edge update index")
+
+	// We'll need to properly prune all the outdated entries within the edge
+	// update index. To do so, we'll gather all of the existing policies
+	// within the graph to re-populate them later on.
+	var edgeKeys [][]byte
+	err := edges.ForEach(func(edgeKey, edgePolicyBytes []byte) error {
+		// All valid entries are indexed by a public key (33 bytes)
+		// followed by a channel ID (8 bytes), so we'll skip any entries
+		// with keys that do not match this.
+		if len(edgeKey) != 33+8 {
+			return nil
+		}
+
+		edgeKeys = append(edgeKeys, edgeKey)
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("unable to gather existing edge policies: %v",
+			err)
+	}
+
+	// With the existing edge policies gathered, we'll recreate the index
+	// and populate it with the correct entries.
+	oldNumEntries := edgeUpdateIndex.Stats().KeyN
+	if err := edges.DeleteBucket(edgeUpdateIndexBucket); err != nil {
+		return fmt.Errorf("unable to remove existing edge update "+
+			"index: %v", err)
+	}
+	edgeUpdateIndex, err = edges.CreateBucketIfNotExists(
+		edgeUpdateIndexBucket,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to recreate edge update index: %v",
+			err)
+	}
+
+	// For each edge key, we'll retrieve the policy, deserialize it, and
+	// re-add it to the different buckets. By doing so, we'll ensure that
+	// all existing edge policies are serialized correctly within their
+	// respective buckets and that the correct entries are populated within
+	// the edge update index.
+	for _, edgeKey := range edgeKeys {
+		edgePolicyBytes := edges.Get(edgeKey)
+
+		// Skip any entries with unknown policies as there will not be
+		// any entries for them in the edge update index.
+		if bytes.Equal(edgePolicyBytes[:], unknownPolicy) {
+			continue
+		}
+
+		edgePolicy, err := deserializeChanEdgePolicy(
```

Comment:
It was pointed out by @joostjager that this might be dangerous if the serialization format changes in the future. Let's say we in DB version 8 we reorder how the channel edge policy is serialized, with a corresponding migration. Now a user running a db upgrade from 5 to 6 will still call into the same method, and this might fail because the DB format doesn't fit what the method expects.

Maybe we should also version the serialization methods, or keep all serialization needed in the migration itself. An alternative would be a test that applies all serializations to catch if something like this happens.